zrml/orderbook/src/lib.rs

```diff
@@ -164,181 +178,274 @@ mod pallet {
+
+            Ok(())
+        }
+
+        /// Charge the external fees from `taker` and return the adjusted maker fill.
+        fn charge_external_fees(
+            order_data: &OrderOf<T>,
+            base_asset: AssetOf<T>,
+            maker_fill: BalanceOf<T>,
+            taker: &AccountIdOf<T>,
+            taker_fill: BalanceOf<T>,
+        ) -> Result<BalanceOf<T>, DispatchError> {
+            let maker_asset_is_base = order_data.maker_asset == base_asset;
+            let base_asset_fill = if maker_asset_is_base {
+                taker_fill
+            } else {
+                debug_assert!(order_data.taker_asset == base_asset);
+                maker_fill
+            };
+            let fee_amount = T::ExternalFees::distribute(
+                order_data.market_id,
+                base_asset,
+                taker,
+                base_asset_fill,
+            );
+            if maker_asset_is_base {
+                // maker_fill is the amount what the maker wants to have (outcome asset from taker)
+                // do not charge fees from outcome assets, but rather from the base asset
+                Ok(maker_fill)
+            } else {
+                // accounting fees from the taker,
+                // who is responsible to pay the base asset minus fees to the maker.
+                Ok(maker_fill.checked_sub_res(&fee_amount)?)
+            }
+        }
+
+        fn do_fill_order(
+            order_id: OrderId,
+            taker: AccountIdOf<T>,
+            maker_partial_fill: Option<BalanceOf<T>>,
+        ) -> DispatchResult {
             let mut order_data = <Orders<T>>::get(order_id).ok_or(Error::<T>::OrderDoesNotExist)?;
             let market = T::MarketCommons::market(&order_data.market_id)?;
-            ensure!(market.scoring_rule == ScoringRule::Orderbook, Error::<T>::InvalidScoringRule);
+            debug_assert!(
+                market.scoring_rule == ScoringRule::Orderbook,
+                "The call to place_order already ensured the scoring rule order book."
+            );
             ensure!(market.status == MarketStatus::Active, Error::<T>::MarketIsNotActive);
             let base_asset = market.base_asset;
 
-            let makers_requested_total = match order_data.side {
-                OrderSide::Bid => order_data.outcome_asset_amount,
-                OrderSide::Ask => order_data.base_asset_amount,
-            };
-            let maker_fill = maker_partial_fill.unwrap_or(makers_requested_total);
-            ensure!(!maker_fill.is_zero(), Error::<T>::AmountIsZero);
-            ensure!(maker_fill <= makers_requested_total, Error::<T>::AmountTooHighForOrder);
+            let maker_fill = maker_partial_fill.unwrap_or(order_data.taker_amount);
+            ensure!(
+                maker_fill >= T::AssetManager::minimum_balance(order_data.taker_asset),
+                Error::<T>::BelowMinimumBalance
+            );
```

Comment:
Does it mean an order gets stuck once the remaining balance the maker wants is below the minimum balance? In that case I recommend moving this check to when the actual transferal of the funds happens, because in most cases the maker will already have some balance for the requested token and fulfill this requirement.

Reply:
It just means that a taker, who uses `fill_order` is not allowed to partially fill an order to the extend that there is an amount missing that is below the existential deposit. We ensure somewhere else in the code that the `taker_amount` stays above the existential deposit.

Reply:
[Here](https://github.com/zeitgeistpm/zeitgeist/blob/7052aefccd264f15d749405764777e9f2e303386/zrml/orderbook/src/lib.rs#L493-L496) we check that `taker_amount` is above the existential deposit. [Here](https://github.com/zeitgeistpm/zeitgeist/blob/7052aefccd264f15d749405764777e9f2e303386/zrml/orderbook/src/lib.rs#L311) we ensure the `taker_amount` is also above the existential deposit in a (partial) fill.